clang/runtime/dpct-rt/include/dpct/math.hpp

```diff
@@ -2055,6 +2057,277 @@ class joint_matrix {
   matrix_accessor x;
   const size_t num_elements;
 };
+
+/// Multiplies 2 8x4 & 4x8 matrices and accumulates the result to a 8x8 b16
+/// matrix
+/// Requires the sub-group size of kernel calling this function to be 32
+/// \tparam [in] MulType The type of the multiplication result
+/// \tparam [in] ABType The type of the input matrices
+/// \tparam [in] CDType The type of the output matrix
+/// \tparam [in] ItemT The type of the sycl::nd_item index space class
+/// \param [in] d0 The 1st element to be written to the output D matrix
+/// \param [in] d1 The 2nd element to be written to the output D matrix
+/// \param [in] d2 The 3rd element to be written to the output D matrix
+/// \param [in] d3 The 4th element to be written to the output D matrix
+/// \param [in] a0 The 1st element from A matrix to be multiplied with B matrix
+/// \param [in] a1 The 2nd element from A matrix to be multiplied with B matrix
+/// \param [in] b0 The 1st element from B matrix to be multiplied with A matrix
+/// \param [in] b1 The 2nd element from B matrix to be multiplied with A matrix
+/// \param [in] c0 The 1st element from C matrix to be added with d0
+/// \param [in] c1 The 2nd element from C matrix to be added with d1
+/// \param [in] c2 The 3rd element from C matrix to be added with d2
+/// \param [in] c3 The 4th element from C matrix to be added with d3
+/// \param [in] item The sycl::nd_item index space class
+template <typename MulType, typename ABType, typename CDType, typename ItemT>
+void mma(CDType *d0, CDType *d1, CDType *d2, CDType *d3, ABType a0, ABType a1,
+         ABType b0, ABType b1, CDType c0, CDType c1, CDType c2, CDType c3,
+         const ItemT &item) {
```

Comment:
How about designing the interface like:
```
template <typename MulType, typename ABType, typename CDType, typename ItemT, unsigned m, unsigned n, unsigned k>
void mma(CDType *d, const ABType* a,  const ABType* b, const CDType* c, const ItemT &item);
```
Then we can merge the impls of different shapes in a single function. And in that function, we can use `if constexpr()` to set different branches for different shapes.

Reply:
```
sturct dpct_int4;
template <typename MulType, typename ABType, typename CDType, typename ItemT,
          unsigned m, unsigned n, unsigned k>
void mma(volatile void *d, const void * a, const void * b, const void * c, const ItemT &item);
```

Reply:
Please focus on this shape first:
```
autoGPTQ:         

mma.sync.aligned.m16n8k16.row.col.f32.f16.f16.f32
 
llama.cpp:

./llama.cpp/ggml/src/ggml-cuda/mma.cuh:175:        asm("mma.sync.aligned.m16n8k16.row.col.s32.s8.s8.s32 {%0, %1, %2, %3}, {%4, %5}, {%6}, {%0, %1, %2, %3};"
 
```